4.x connection cache

ADmad opened the pull request:
Similar to #13276, this change allows connection to now use injectable simple cache instance.

ADmad committed:
Add ConnectionInterface::setCacher()/getCacher()

ADmad committed:
Use connection's cache instance instead of directly fetching from cache pool.

lorenzo commented on src/Database/Connection.php:
```diff
@@ -790,6 +799,33 @@ public function cacheMetadata($cache): void
+                $configName = '_cake_model_';
+            }
+
+            $this->cacher = Cache::pool($configName);
```
I think that   we should somehow not force this dependency, and maybe provide a null cacher implementation instead of defaulting in this way.

ADmad replied:
If we remove this then we will need to come up with another way to set default cacher for cake apps.

Currently we just call `ConnectionManager` in bootstrap to setup db config arrays and the rest happens "magically".

ADmad replied:
I can make `getCacher()` return null if no cacher is explicitly set and move the fetching to default cache using `Cache` back into `CachedCollection`.

markstory replied:
Would it be better if we raised an error if `cakephp/cache` isn't available?

ADmad replied:
Done. I also removed `cakehp/cache` as dependency.

ADmad committed:
Fix SchemaCacheShell tests

ADmad committed:
Merge branch '4.x' into 4.x-connection-cache

ADmad committed:
Remove cakephp/cache as dependency of database package.

Throw exception if cacher for connection is not explicitly set and Cache class is not available.

ADmad commented:
The tests seem to be failing because same connection instance is used across tests and it's state is not reset.

I cloned the connection instance in few tests and locally I don't get any errors when running phpunit for the `TestCase/Database` folder or `SchemaCacheShellTest.php` separately. But still can't figure out what's causing failure on travis.

markstory commented on src/Database/composer.json:
```diff
@@ -25,7 +25,6 @@
     },
     "require": {
         "php": ">=7.2.0",
-        "cakephp/cache": "^4.0",
```
Should we add the psr/simple-cache package instead?

markstory commented on tests/TestCase/Shell/SchemaCacheShellTest.php:
```diff
@@ -65,8 +67,23 @@ public function tearDown(): void
+        unset($this->connection);
+    }
+
+    protected function getShell()
```
You and I will conflict on this.

ADmad replied:
Yeah :disappointed:

markstory reviewed (approved):
Looks good but we need to update the upgrade guide as there are a few breaking changes here. I don't think they will have significant impacts in applications but still good to document the changes.

ADmad committed:
Add psr/simple-cache as dependency for datasource package.

ADmad committed:
Merge branch '4.x' into 4.x-connection-cache

ADmad committed:
Unset cacher if cache config name is updated.

ADmad committed:
Merge branch '4.x' into 4.x-connection-cache

ADmad committed:
Merge branch '4.x' into 4.x-connection-cache

ADmad commented:
I fixed the issue that was causing build failures earlier.

But now the testsuite crashes for MySql and Postgres with duplicate foreign key contraint error. It's the same error that has been causing Appveyor/MSSQL build to fail since a long time. So surely not caused by my changes.

ADmad committed:
Remove unneeded connection cloning in tests.

ADmad committed:
Use mock to avoid throwing error to STDERR

markstory committed:
Remove redundant test logic

This code was doubled up and was causing the test suite to fail.

markstory committed:
Merge branch '4.x' into 4.x-connection-cache

markstory committed:
Fix grammar

markstory committed:
Fix incorrect test.

This test stub no longer loads ArticlesTags it can't be used for
assertions

markstory committed:
Fix failing tests hopefully.

One of the connection tests was enabling metadata caching which was
causing tests to fail elsewhere.

I've switched Connection::disableConstraints() to use `finally` as it is
simpler and functionally equivalent.

markstory committed:
Don't rely on global state setup.

markstory committed:
Remove unused import

markstory committed:
Expand potential types to isCallable()

Callers could have ExpressionInterface objects they need to test.

markstory committed:
Code downstream of exists() doesn't support ArrayAccess

markstory committed:
Update baseline.

psalm errors have changed.

markstory committed:
Fix psalm better
